### PR TITLE
ensure CRD in framework boot

### PR DESCRIPTION
**Pull request by xh3b4sd**

### CODE REVIEW

File: framework/framework.go [hunk 1]
```diff
@@ -650,10 +663,23 @@ func (f *Framework) bootWithError(ctx context.Context) error {
...
+
+		// TODO collect metrics
+	}
+
 	f.logger.LogCtx(ctx, "debug", "starting list/watch")
 
-	deleteChan, updateChan, errChan := f.informer.Watch(context.TODO())
-	f.ProcessEvents(context.TODO(), deleteChan, updateChan, errChan)
+	deleteChan, updateChan, errChan := f.informer.Watch(ctx)
+	f.ProcessEvents(ctx, deleteChan, updateChan, errChan)
```
Comment (xh3b4sd) [inline framework/framework.go]: Boyscouting the shit out of it.

File: framework/framework.go [hunk 2]
```diff
@@ -650,10 +663,23 @@ func (f *Framework) bootWithError(ctx context.Context) error {
...
+		f.logger.LogCtx(ctx, "debug", "ensuring custom resource definition exists")
+
+		err := f.crdClient.Ensure(ctx, f.crd, f.backOffFactory())
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		f.logger.LogCtx(ctx, "debug", "ensured custom resource definition")
+
+		// TODO collect metrics
```
Comment (xh3b4sd) [inline framework/framework.go]: Comes separately.

File: framework/framework.go [hunk 3]
```diff
@@ -25,7 +27,9 @@ import (
 type Config struct {
 	// Dependencies.
 
-	Informer informer.Interface
+	CRD       *crd.CRD
+	CRDClient *crdclient.CRDClient
```
Comment (xh3b4sd) [inline framework/framework.go]: Note my other PR where this got moved. Please see https://github.com/giantswarm/operatorkit/pull/124 first so we can merge and align here.

**Comment by xh3b4sd** [comment]
I think the good thing on this is that you just throw everything into the framework and it sorts the rest. When each operator would need to ensure the CRD on its own you would need to wire this in your operator to be able to ensure when you "boot". The alternative would be to ensure in a constructor of an operator which would be worse.

**Comment by MarcelMue** [comment]
Just for a quick recap for me:
This PR improves that if you have a `CRD` or a `CRDCLient`, then you always have to have both after this PR. This does not force every operator to have a CRD, more just if there is one, then there should also be a client. And the other way around. 
Am I getting this right?

**Comment by xh3b4sd** [comment]
Yes.

**Pull request merged**